addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/HistoryProvider.java

```diff
@@ -19,13 +19,84 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.extension.history.ExtensionHistory;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
 import org.parosproxy.paros.network.HttpMessage;
 import org.zaproxy.zap.authentication.AuthenticationHelper;
 
 /** A very thin layer on top of the History functionality, to make testing easier. */
 public class HistoryProvider {
 
+    private static final int MAX_NUM_RECORDS_TO_CHECK = 200;
+
+    private static final Logger LOGGER = LogManager.getLogger(HistoryProvider.class);
+
+    private ExtensionHistory extHist;
+
+    private ExtensionHistory getExtHistory() {
+        if (extHist == null) {
+            extHist = AuthUtils.getExtension(ExtensionHistory.class);
+        }
+        return extHist;
+    }
+
     public void addAuthMessageToHistory(HttpMessage msg) {
         AuthenticationHelper.addAuthMessageToHistory(msg);
     }
+
+    public HttpMessage getHttpMessage(int historyId)
+            throws HttpMalformedHeaderException, DatabaseException {
+        HistoryReference hr = getExtHistory().getHistoryReference(historyId);
+        if (hr != null) {
+            return hr.getHttpMessage();
+        }
+        return null;
+    }
+
+    public int getLastHistoryId() {
+        return getExtHistory().getLastHistoryId();
+    }
+
+    public SessionManagementRequestDetails findSessionTokenSource(String token, int firstId) {
+        int lastId = getLastHistoryId();
+        if (firstId == -1) {
+            firstId = Math.max(0, lastId - MAX_NUM_RECORDS_TO_CHECK);
+        }
+
+        LOGGER.debug("Searching for session token from {} down to {} ", lastId, firstId);
+
+        for (int i = lastId; i >= firstId; i--) {
+            try {
+                HttpMessage msg = getHttpMessage(i);
+                if (msg == null) {
+                    continue;
+                }
+                Optional<SessionToken> es =
+                        AuthUtils.getAllTokens(msg, false).values().stream()
+                                .filter(v -> v.getValue().equals(token))
+                                .findFirst();
+                if (es.isPresent()) {
+                    AuthUtils.incStatsCounter(
+                            msg.getRequestHeader().getURI(),
+                            AuthUtils.AUTH_SESSION_TOKEN_STATS_PREFIX + es.get().getKey());
+                    List<SessionToken> tokens = new ArrayList<>();
+                    tokens.add(
+                            new SessionToken(
+                                    es.get().getSource(), es.get().getKey(), es.get().getValue()));
+                    return new SessionManagementRequestDetails(msg, tokens, Alert.CONFIDENCE_HIGH);
+                }
+            } catch (Exception e) {
```

Comment:
Is there a reason this can't catch the relevant exceptions?

Reply:
We could, but we'd do exactly the same thing, so how does theat benefit us?

Reply:
I guess it doesn't matter ultimately. The main benefit would be if something unanticipated happened I guess.